cartofacade/carto_facade_queue.go

```diff
@@ -0,0 +1,151 @@
+	Start
+	// Stop can be used to represent the viam_carto_stop call in c.
+	Stop
+	// Terminate can be used to represent the viam_carto_terminate in c.
+	Terminate
+	// AddSensorReading can be used to represent the viam_carto_add_sensor_reading in c.
+	AddSensorReading
+	// Position can be used to represent the viam_carto_get_position call in c.
+	Position
+	// InternalState can be used to represent the viam_carto_get_internal_state call in c.
+	InternalState
+	// PointCloudMap can be used to represent the viam_carto_get_point_cloud_map in c.
+	PointCloudMap
+)
+
+// InputType defines the type being provided as input to the work.
+type InputType int64
+
+const (
+	// Reading represents a lidar reading input into c funcs.
+	Reading InputType = iota
+	// Timestamp represents the timestamp input into c funcs.
+	Timestamp
+)
+
+// WorkItem defines one piece of work that can be put on the queue.
+type WorkItem struct {
+	Result   chan interface{}
+	workType WorkType
+	inputs   map[InputType]interface{}
+}
+
+// DoWork provides the logic to call the correct cgo functions with the correct input.
+func (w *WorkItem) DoWork(
+	q *Queue,
+) (interface{}, error) {
+	switch w.workType {
+	case Initialize:
+		return New(q.CartoConfig, q.CartoAlgoConfig, *q.CartoLib)
+	case Start:
+		return nil, q.Carto.Start()
+	case Stop:
+		return nil, q.Carto.Stop()
+	case Terminate:
+		return q.Carto.Terminate(), nil
+	case AddSensorReading:
+		reading, ok := w.inputs[Reading].([]byte)
+		if !ok {
+			return nil, errors.New("could not cast inputted byte to byte slice")
+		}
+
+		timestamp, ok := w.inputs[Timestamp].(time.Time)
+		if !ok {
+			return nil, errors.New("could not cast inputted timestamp to times.Time")
+		}
+
+		return nil, q.Carto.AddSensorReading(reading, timestamp)
+	case Position:
+		return q.Carto.GetPosition()
+	case InternalState:
+		return q.Carto.GetInternalState()
+	case PointCloudMap:
+		return q.Carto.GetPointCloudMap()
+	}
+	return nil, fmt.Errorf("no worktype found for: %v", w.workType)
+}
+
+// Queue represents a queue to consume work from and enforce one call into C at a time.
+type Queue struct {
+	WorkChannel     chan WorkItem
+	CartoLib        *CartoLib
+	Carto           Carto
+	CartoConfig     CartoConfig
+	CartoAlgoConfig CartoAlgoConfig
+}
+
+// NewQueue instantiates the Queue struct.
+func NewQueue(cartoLib *CartoLib, cartoCfg CartoConfig, cartoAlgoCfg CartoAlgoConfig) Queue {
+	return Queue{
+		WorkChannel:     make(chan WorkItem),
+		Carto:           Carto{},
+		CartoLib:        cartoLib,
+		CartoConfig:     cartoCfg,
+		CartoAlgoConfig: cartoAlgoCfg,
+	}
+}
+
+// HandleIncomingRequest puts incoming requests on the queue and consumes from queue.
+func (q *Queue) HandleIncomingRequest(ctx context.Context, workType WorkType, inputs map[InputType]interface{}) interface{} {
+	// Question: What is reasonable timeout period?
+	ctx, cancel := context.WithTimeout(ctx, 15*time.Second)
```

Comment:
If we are providing it as an input, why do we need to define it as a constant at the top?

Reply:
good point, it will be provided as a constant in whatever package calls this function. So nix the comment about moving it to a constant in this package